File: src/rex.widget/js/lib/KeyPath.js
/**
 * @copyright 2015, Prometheus Research, LLC
 */

import {isArray} from './lang';

export function get(keyPath, obj) {
  keyPath = normalize(keyPath);
  for (let i = 0, len = keyPath.length; i < len; i++) {
    if (obj == null) {
      return obj;
    }
    obj = obj[keyPath[i]];
  }
  return obj;
}

export function equals(a, b) {
  a = normalize(a);
  b = normalize(b);
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (a[i] != b[i]) {
      return false;
    }
  }
  return true;
}

export function normalize(keyPath) {
  if (keyPath === null) {
    return [];
  }
  if (!isArray(keyPath)) {
    keyPath = keyPath.split('.').filter(Boolean);
  }
  return keyPath;
}

File: src/rex.widget/src/rex/widget/keypath.py
"""

    rex.widget.keypath
    ==================

    :copyright: 2015, Prometheus Research, LLC

"""

from rex.core import Error, Validate
from rex.core import OneOfVal, SeqVal, StrVal, IntVal

__all__ = ('KeyPathVal',)


class KeyPathVal(Validate):

    _validate = SeqVal(OneOfVal(IntVal(), StrVal()))
    _validate_with_shortcut = OneOfVal(_validate, StrVal(), IntVal())

    def __init__(self, allow_empty=False):
        self.allow_empty = allow_empty

    def __call__(self, value):
        value = self._validate_with_shortcut(value)
        if isinstance(value, str):
            if '/' in value:
                value = value[:value.find('/')]
            if '.' in value:
                return self([v for v in value.split('.') if v])
            elif value == '':
                return self([])
            else:
                return self([value])
        elif isinstance(value, int):
            return [value]
        else:
            if not self.allow_empty and len(value) == 0:
                raise Error('cannot be empty')
            return [try_to_int(v) for v in value]

    @staticmethod
    def to_string(keypath):
        return '.'.join(str(v) for v in keypath)


def try_to_int(v):
    try:
        return int(v)
    except ValueError:
        return v
